File: app/services/view_service.py
"""Query notes for a structured view.

Filters, sorts and groups over the readable notes. A locked layer contributes
nothing — the candidate set is ``NoteService.list_notes()``, which is readable-only,
so a view can never surface a locked note's title or property by construction rather
than by a filter that could be forgotten.

Everything is computed from the live notes on each call. That is the whole point of
"Markdown stays the source of truth": the view is a lens, not a store.
"""

from __future__ import annotations

from datetime import date, datetime, timezone

from app.domain.note import Note
from app.domain.views import (
    ViewConfig,
    ViewFilter,
    ViewGroup,
    ViewResult,
    ViewRow,
)
from app.services.note_service import NoteService
from app.services.workspace_service import WorkspaceService

SNIPPET_LENGTH = 140
UNGROUPED = "\x00ungrouped"


class ViewService:
    def __init__(self, workspace: WorkspaceService, notes: NoteService) -> None:
        self._workspace = workspace
        self._notes = notes

    def run(self, config: ViewConfig) -> ViewResult:
        layer_ids = config.layer_ids or None
        notes = self._notes.list_notes(layer_ids)

        layer_names = {layer.id: layer.display_name for layer in self._workspace.descriptor.layers}
        private_layers = {
            layer.id for layer in self._workspace.descriptor.layers if layer.visibility == "private"
        }

        rows: list[ViewRow] = []
        for note in notes:
            if config.folder_scope and not note.metadata.folder_path.startswith(
                config.folder_scope
            ):
                continue
            if not all(self._matches(note, f) for f in config.filters):
                continue
            rows.append(self._row(note, layer_names, private_layers))

        rows = self._sort(rows, config)
        groups = self._group(rows, config)

        available = sorted(
            {key for row in rows for key in row.properties}
            | {"title", "tags", "folder", "created", "updated"}
        )

        return ViewResult(
            config=config,
            rows=rows,
            groups=groups,
            total=len(rows),
            available_properties=available,
            locked_layers_excluded=len(self._workspace.locked_layers()),
        )

    # -- rows ----------------------------------------------------------------

    def _row(self, note: Note, layer_names: dict[str, str], private_layers: set[str]) -> ViewRow:
        meta = note.metadata
        return ViewRow(
            object_id=meta.id,
            layer_id=meta.layer_id,
            layer_name=layer_names.get(meta.layer_id, "Unknown"),
            is_private=meta.layer_id in private_layers,
            title=meta.title,
            folder_path=meta.folder_path,
            tags=list(meta.tags),
            properties={
                key: _stringify(value)
                for key, value in meta.properties.items()
                if key not in ("tags", "title")
            },
            created_at=meta.created_at,
            updated_at=meta.updated_at,
            snippet=note.content.strip()[:SNIPPET_LENGTH].replace("\n", " "),
        )

    def _field(self, row: ViewRow, field: str) -> str:
        if field == "title":
            return row.title
        if field == "tags":
            return ", ".join(row.tags)
        if field == "folder":
            return row.folder_path
        if field == "layer":
            return row.layer_name
        if field == "created":
            return row.created_at
        if field == "updated":
            return row.updated_at
        return row.properties.get(field, "")

    # -- filtering -----------------------------------------------------------

    def _matches(self, note: Note, view_filter: ViewFilter) -> bool:
        value = self._note_field(note, view_filter.field)
        target = view_filter.value
        op = view_filter.operator

        if op == "is_empty":
            return not value
        if op == "is_not_empty":
            return bool(value)
        if op == "equals":
            return value.lower() == target.lower()
        if op == "not_equals":
            return value.lower() != target.lower()
        if op == "contains":
            return target.lower() in value.lower()
        if op == "not_contains":
            return target.lower() not in value.lower()
        if op == "in":
            options = [item.strip().lower() for item in target.split(",")]
            return value.lower() in options
        if op in ("greater_than", "less_than"):
            return self._numeric_compare(value, target, op)
        if op in ("before", "after"):
            return self._date_compare(value, target, op)
        return False

    def _note_field(self, note: Note, field: str) -> str:
        meta = note.metadata
        if field == "title":
            return meta.title
        if field == "tags":
            return ", ".join(meta.tags)
        if field == "folder":
            return meta.folder_path
        if field == "layer":
            return meta.layer_id
        if field == "created":
            return meta.created_at
        if field == "updated":
            return meta.updated_at
        return _stringify(meta.properties.get(field, ""))

    @staticmethod
    def _numeric_compare(value: str, target: str, op: str) -> bool:
        try:
            left, right = float(value), float(target)
        except ValueError:
            return False
        return left > right if op == "greater_than" else left < right

    @staticmethod
    def _date_compare(value: str, target: str, op: str) -> bool:
        left = _parse_date(value)
        right = _parse_date(target)
        if left is None or right is None:
            return False
        return left < right if op == "before" else left > right

    # -- sorting and grouping ------------------------------------------------

    def _sort(self, rows: list[ViewRow], config: ViewConfig) -> list[ViewRow]:
        for sort in reversed(config.sort):
            rows = sorted(
                rows,
                key=lambda row: _sort_key(self._field(row, sort.field)),
                reverse=(sort.direction == "desc"),
            )
        if not config.sort:
            rows = sorted(rows, key=lambda row: row.title.lower())
        return rows

    def _group(self, rows: list[ViewRow], config: ViewConfig) -> list[ViewGroup]:
        if not config.group_by:
            return []

        buckets: dict[str, list[ViewRow]] = {}
        order: list[str] = []
        for row in rows:
            key = self._field(row, config.group_by) or UNGROUPED
            if key not in buckets:
                buckets[key] = []
                order.append(key)
            buckets[key].append(row)

        return [
            ViewGroup(
                key="" if key == UNGROUPED else key,
                label="No value" if key == UNGROUPED else key,
                rows=buckets[key],
            )
            for key in order
        ]


def _stringify(value: object) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, list):
        return ", ".join(str(item) for item in value)
    return str(value) if value is not None else ""


def _sort_key(value: str) -> tuple[int, float, str]:
    """Numbers sort numerically, everything else lexically, empties last."""
    if not value:
        return (2, 0.0, "")
    try:
        return (0, float(value), "")
    except ValueError:
        return (1, 0.0, value.lower())


def _parse_date(value: str) -> datetime | None:
    if not value:
        return None
    try:
        parsed = datetime.fromisoformat(value)
    except ValueError:
        try:
            parsed = datetime.combine(date.fromisoformat(value[:10]), datetime.min.time())
        except ValueError:
            return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed

File: dotnet/src/Strata.Services/ViewService.cs
using System.Globalization;
using System.Text.Json.Nodes;
using Strata.Core.Layers;
using Strata.Core.Notes;
using Strata.Core.Views;

namespace Strata.Services;

/// <summary>
/// Query notes for a structured view — port of
/// <c>app/services/view_service.py</c>.
/// </summary>
/// <remarks>
/// <para>
/// Filters, sorts and groups over the readable notes. A locked layer contributes
/// nothing — the candidate set is <see cref="INoteSource.ListNotes"/>, which is
/// readable-only, so a view can never surface a locked note's title or property by
/// construction rather than by a filter that could be forgotten.
/// </para>
/// <para>
/// Everything is computed from the live notes on each call. That is the whole point
/// of "Markdown stays the source of truth": the view is a lens, not a store.
/// </para>
/// </remarks>
public sealed class ViewService(WorkspaceService workspace, INoteSource notes)
{
    public const int SnippetLength = 140;

    /// <summary>Sentinel for the ungrouped bucket; cannot collide with a real value.</summary>
    private const string Ungrouped = "\0ungrouped";

    public ViewResult Run(ViewConfig config)
    {
        var layerIds = config.LayerIds.Count > 0 ? config.LayerIds : null;
        var candidates = notes.ListNotes(layerIds);

        var layerNames = workspace.Descriptor.Layers.ToDictionary(
            layer => layer.Id, layer => layer.DisplayName, StringComparer.Ordinal);
        var privateLayers = workspace.Descriptor.Layers
            .Where(layer => layer.Visibility == LayerVisibility.Private)
            .Select(layer => layer.Id)
            .ToHashSet(StringComparer.Ordinal);

        var rows = new List<ViewRow>();
        foreach (var note in candidates)
        {
            if (config.FolderScope.Length > 0
                && !note.Metadata.FolderPath.StartsWith(config.FolderScope, StringComparison.Ordinal))
            {
                continue;
            }

            if (!config.Filters.All(filter => Matches(note, filter)))
            {
                continue;
            }

            rows.Add(Row(note, layerNames, privateLayers));
        }

        rows = Sort(rows, config);
        var groups = Group(rows, config);

        var available = rows.SelectMany(row => row.Properties.Keys)
            .Concat(["title", "tags", "folder", "created", "updated"])
            .Distinct(StringComparer.Ordinal)
            .Order(StringComparer.Ordinal)
            .ToList();

        return new ViewResult
        {
            Config = config,
            Rows = rows,
            Groups = groups,
            Total = rows.Count,
            AvailableProperties = available,
            LockedLayersExcluded = workspace.LockedLayers().Count,
        };
    }

    // -- rows ----------------------------------------------------------------

    private static ViewRow Row(Note note, Dictionary<string, string> layerNames, HashSet<string> privateLayers)
    {
        var meta = note.Metadata;
        return new ViewRow
        {
            ObjectId = meta.Id,
            LayerId = meta.LayerId,
            LayerName = layerNames.GetValueOrDefault(meta.LayerId, "Unknown"),
            IsPrivate = privateLayers.Contains(meta.LayerId),
            Title = meta.Title,
            FolderPath = meta.FolderPath,
            Tags = [.. meta.Tags],
            Properties = meta.Properties
                .Where(pair => pair.Key is not ("tags" or "title"))
                .ToDictionary(pair => pair.Key, pair => Stringify(pair.Value), StringComparer.Ordinal),
            CreatedAt = meta.CreatedAt,
            UpdatedAt = meta.UpdatedAt,
            Snippet = Snippet(note.Content),
        };
    }

    private static string Snippet(string content)
    {
        var trimmed = content.Trim();
        var clipped = trimmed.Length > SnippetLength ? trimmed[..SnippetLength] : trimmed;
        return clipped.Replace("\n", " ", StringComparison.Ordinal);
    }

    private static string Field(ViewRow row, string field) => field switch
    {
        "title" => row.Title,
        "tags" => string.Join(", ", row.Tags),
        "folder" => row.FolderPath,
        "layer" => row.LayerName,
        "created" => row.CreatedAt,
        "updated" => row.UpdatedAt,
        _ => row.Properties.GetValueOrDefault(field, string.Empty),
    };

    /// <summary>
    /// The same pseudo-fields as <see cref="Field"/>, but resolved on the note.
    /// </summary>
    /// <remarks>
    /// Note the deliberate difference from the row version: filtering on "layer"
    /// matches the layer <em>id</em>, while a row displays the layer's name. Both
    /// behaviours are inherited from the Python original.
    /// </remarks>
    private static string NoteField(Note note, string field)
    {
        var meta = note.Metadata;
        return field switch
        {
            "title" => meta.Title,
            "tags" => string.Join(", ", meta.Tags),
            "folder" => meta.FolderPath,
            "layer" => meta.LayerId,
            "created" => meta.CreatedAt,
            "updated" => meta.UpdatedAt,
            _ => Stringify(meta.Properties.TryGetPropertyValue(field, out var value) ? value : null),
        };
    }

    // -- filtering -----------------------------------------------------------

    private static bool Matches(Note note, ViewFilter filter)
    {
        var value = NoteField(note, filter.Field);
        var target = filter.Value;

        return filter.Operator switch
        {
            FilterOperator.IsEmpty => value.Length == 0,
            FilterOperator.IsNotEmpty => value.Length > 0,
            FilterOperator.Equals => Same(value, target),
            FilterOperator.NotEquals => !Same(value, target),
            FilterOperator.Contains => value.Contains(target, StringComparison.OrdinalIgnoreCase),
            FilterOperator.NotContains => !value.Contains(target, StringComparison.OrdinalIgnoreCase),
            FilterOperator.In => target.Split(',')
                .Select(item => item.Trim())
                .Any(item => Same(value, item)),
            FilterOperator.GreaterThan or FilterOperator.LessThan
                => NumericCompare(value, target, filter.Operator),
            FilterOperator.Before or FilterOperator.After
                => DateCompare(value, target, filter.Operator),
            _ => false,
        };
    }

    private static bool Same(string left, string right)
        => string.Equals(left, right, StringComparison.OrdinalIgnoreCase);

    private static bool NumericCompare(string value, string target, FilterOperator op)
    {
        if (!TryNumber(value, out var left) || !TryNumber(target, out var right))
        {
            return false;
        }

        return op == FilterOperator.GreaterThan ? left > right : left < right;
    }

    private static bool DateCompare(string value, string target, FilterOperator op)
    {
        if (ParseDate(value) is not { } left || ParseDate(target) is not { } right)
        {
            return false;
        }

        return op == FilterOperator.Before ? left < right : left > right;
    }

    // -- sorting and grouping ------------------------------------------------

    private static List<ViewRow> Sort(List<ViewRow> rows, ViewConfig config)
    {
        if (config.Sort.Count == 0)
        {
            return rows.OrderBy(row => row.Title.ToLowerInvariant(), StringComparer.Ordinal).ToList();
        }

        // Apply each key in reverse with a stable sort, so the first key wins —
        // the same technique the Python original uses.
        foreach (var sort in config.Sort.Reverse())
        {
            rows = sort.Direction == SortDirection.Desc
                ? rows.OrderByDescending(row => SortKey(Field(row, sort.Field))).ToList()
                : rows.OrderBy(row => SortKey(Field(row, sort.Field))).ToList();
        }

        return rows;
    }

    private static List<ViewGroup> Group(List<ViewRow> rows, ViewConfig config)
    {
        if (config.GroupBy.Length == 0)
        {
            return [];
        }

        var buckets = new Dictionary<string, List<ViewRow>>(StringComparer.Ordinal);
        var order = new List<string>();

        foreach (var row in rows)
        {
            var value = Field(row, config.GroupBy);
            var key = value.Length > 0 ? value : Ungrouped;
            if (!buckets.TryGetValue(key, out var bucket))
            {
                bucket = [];
                buckets[key] = bucket;
                order.Add(key);
            }

            bucket.Add(row);
        }

        return order.Select(key => new ViewGroup
        {
            Key = key == Ungrouped ? string.Empty : key,
            Label = key == Ungrouped ? "No value" : key,
            Rows = buckets[key],
        }).ToList();
    }

    // -- coercion ------------------------------------------------------------

    private static string Stringify(JsonNode? value)
    {
        if (value is null)
        {
            return string.Empty;
        }

        if (value is JsonArray array)
        {
            return string.Join(", ", array.Select(Stringify));
        }

        if (value is JsonValue scalar)
        {
            if (scalar.TryGetValue<bool>(out var flag))
            {
                return flag ? "true" : "false";
            }

            if (scalar.TryGetValue<string>(out var text))
            {
                return text;
            }
        }

        return value.ToJsonString().Trim('"');
    }

    private static bool TryNumber(string value, out double number)
        => double.TryParse(value, NumberStyles.Float, CultureInfo.InvariantCulture, out number);

    /// <summary>Numbers sort numerically, everything else lexically, empties last.</summary>
    private static (int Bucket, double Number, string Text) SortKey(string value)
    {
        if (value.Length == 0)
        {
            return (2, 0.0, string.Empty);
        }

        return TryNumber(value, out var number)
            ? (0, number, string.Empty)
            : (1, 0.0, value.ToLowerInvariant());
    }

    private static DateTimeOffset? ParseDate(string value)
    {
        if (value.Length == 0)
        {
            return null;
        }

        if (DateTimeOffset.TryParse(value, CultureInfo.InvariantCulture,
                DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal, out var parsed))
        {
            return parsed;
        }

        // Fall back to the leading date, matching Python's date.fromisoformat(value[:10]).
        if (value.Length >= 10
            && DateOnly.TryParseExact(value[..10], "yyyy-MM-dd", CultureInfo.InvariantCulture,
                DateTimeStyles.None, out var date))
        {
            return new DateTimeOffset(date.ToDateTime(TimeOnly.MinValue), TimeSpan.Zero);
        }

        return null;
    }
}
